presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateProjectOnValues.java

```diff
@@ -0,0 +1,137 @@
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.planner.plan.Patterns.values;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * This optimizer looks for ProjectNode followed by a ValuesNode and get the ProjectNode Evaluated.
+ * <p/>
+ * Plan before optimizer:
+ * <pre>
+ * ProjectNode (outputVariables)
+ *   - ValuesNode
+ * </pre>
+ * <p/>
+ * Plan after optimizer:
+ * <pre>
+ * ValuesNode (outputVariables)
+ * </pre>
+ */
+public class EvaluateProjectOnValues
+        implements Rule<ProjectNode>
+{
+    private static final Pattern<ProjectNode> PATTERN = project().with(source().matching(values()));
+
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public EvaluateProjectOnValues(FunctionAndTypeManager functionAndTypeManager)
+    {
+        requireNonNull(functionAndTypeManager, "functionManager is null");
+        this.functionAndTypeManager = functionAndTypeManager;
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isEvaluateProjectOnValues(session);
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        Optional<PlanNode> optionalSource = context.getLookup().resolveGroup(projectNode.getSource()).findFirst();
+        if (!optionalSource.isPresent()) {
+            return Result.empty();
+        }
+        ValuesNode source = (ValuesNode) optionalSource.get();
+        Collection<RowExpression> projectRowExpressions = projectNode.getAssignments().getExpressions();
+        List<List<RowExpression>> rows = source.getRows();
+        List<VariableReferenceExpression> valuesOutputVariables = source.getOutputVariables();
+
+        //evaluate ProjectNode
+        List<List<Object>> rowValues = rows.stream().map(row -> {
+            //Prepare to set up variable resolver
+            Map<String, Object> valuesMapForResolver = new HashMap<>();
+            IntStream.range(0, row.size()).forEach(index -> {
+                RowExpression element = row.get(index);
+                RowExpressionInterpreter elementInterpreter = new RowExpressionInterpreter(element,
+                        functionAndTypeManager,
+                        context.getSession().toConnectorSession(),
+                        EVALUATED);
+                valuesMapForResolver.put(valuesOutputVariables.get(index).getName(), elementInterpreter.evaluate());
+            });
+            VariableResolver variableResolver = new Interpreters.LambdaVariableResolver(valuesMapForResolver);
+            //evaluate each row of the ProjectNode
+            List<Object> rowValue = new ArrayList<>();
+            for (RowExpression rowExpression : projectRowExpressions) {
```

Comment:
I would pick a style (functional or imperative) and stick to it rather than mixing the two. For this loop you can instead do `projectRowExpressions.stream().map(...).collect(toImmutableList())`

Reply:
I agree though I think a judicious mix of things is ok.

Reply:
I refactored a little bit. Hope it is better structured.